### PR TITLE
Simplify use of Trilinos-exported Kokkos: Issue #2779

### PR DESCRIPTION
**Pull request by jjwilke**
This address the customer request in #2779. It provides an easy-to-use exported target `Trilinos::kokkos` for those wanting to use targets.  It also completely removes the `kokkos.pc` which is not really correct or supported anymore.

### CODE REVIEW

**Comment by jjwilke** [comment]
Wait on customer testing in https://github.com/kokkos/kokkos/issues/2779 before merging.

**Comment by aprokop** [comment]
Here's what I get right now:
```cmake
IF (NOT TARGET Trilinos::kokkos)                                                                           
  add_library(Trilinos::kokkos INTERFACE IMPORTED)                                                         
  set_target_properties(Trilinos::kokkos PROPERTIES                                                        
    INTERFACE_LINK_LIBRARIES "kokkosalgorithms;kokkoscontainers;kokkoscore"                                
    INTERFACE_COMPILE_FEATURES "cxx_std_11"                                                                
    INTERFACE_COMPILE_OPTIONS ""                                                                           
    INTERFACE_INCLUDE_DIRECTORIES "/home/xap/code/trilinos/master/build_kokkos;/home/xap/code/trilinos/master/build_kokkos/kokkos/algorithms/src;/home/xap/code/trilinos/master/kokkos/algorithms/src;/home/xap/code/trilinos/master/build_kokkos/kokkos/containers/src;/home/xap/code/trilinos/master/kokkos/containers/src;/home/xap/code/trilinos/master/build_kokkos/kokkos/core/src;/home/xap/code/trilinos/master/kokkos/core/src;/home/xap/code/trilinos/master/build_kokkos/kokkos"                                                            
    INTERFACE_LINK_OPTIONS "-fopenmp"                                                                      
  )                                                                                                        
ENDIF()       
```
So
- `INTERFACE_COMPILE_OPTIONS` is empty :(
- `INTERFACE_INCLUDE_DIRECTORIES` is also off
  It erroneously contains source and build directories, but not the install (in standalone installation, it is `"${_IMPORT_PREFIX}/include"`)
- I wonder why this target has both `INTERFACE_LINK_LIBRARIES` and `INTERFACE_LINK_OPTIONS`, while the standalone Kokkos target mixes them both into `INTERFACE_LINK_LIBRARIES`

**Comment by jjwilke** [comment]
I think I fixed the other issues. I'm getting the correct compile options right now. What configure did you use so I can try to reproduce?

**Comment by aprokop** [comment]
[trilinos_configure.txt](https://github.com/kokkos/kokkos/files/4277955/trilinos_configure.txt)
run with `./trilinos_configure -DKokkos_SOURCE_DIR_OVERRIDE:STRING=kokkos`

**Comment by aprokop** [comment]
`INTERFACE_COMPILE_OPTIONS` is still empty for me. And as they are not transitive, I'm not getting them from `kokkoscore` and such.

**Comment by jjwilke** [comment]
Options should be fixed with the latest push.

**Comment by aprokop** [comment]
Better but still not enough:
```
    INTERFACE_COMPILE_OPTIONS "-Xcompiler -fopenmp"                                                                                                                                                                                                                                                                  
```
Missing `-expt-extended-lambda`, for example. Compare to `Kokkos_CXX_FLAGS`:
```
SET(Kokkos_CXX_FLAGS " -pedantic -Wall -Wno-long-long -Wwrite-strings -Wall         -lineinfo         -Xcudafe --diag_suppress=initialization_not_reachable         -Xcudafe --diag_suppress=cc_clobber_ignored  -Xcudafe --diag_suppress=conversion_function_not_usable -Xcudafe --diag_suppress=code_is_unreachable  -std=c++11 -Xcompiler -fopenmp -expt-extended-lambda -arch=sm_61  -O2 -g -DNDEBUG")


```

**Comment by aprokop** [comment]
Also getting errors like
```
nvcc_wrapper  -DBOOST_TEST_DYN_LINK -Itest -I../src -I../src/details -Iinclude -isystem /home/xap/local/opt/trilinos-kokkos/include -isystem /home/xap/local/opt/spack
/opt/spack/linux-ubuntu18.04-broadwell/gcc-7.4.0/openmpi-3.1.4-oetr45yq7vnbcqhiiaalfkpoftuujlzn/include -isystem /home/xap/local/opt/spack/opt/spack/linux-ubuntu18.04-broadwell/gcc-7.4.0/boost-1.68.0-syfratmpslzh7
vpultgho6h54mwh4g34/include -lineinfo -DKOKKOS_ENABLE_PROFILING -O2 -g -DNDEBUG   "-Xcompiler -fopenmp" -fexceptions -pthread -std=c++14 -MD -MT test/CMakeFiles/ArborX_DetailsBatchedQueries.exe.dir/utf_main.cpp.o -MF test/CMakeFiles/ArborX_DetailsBatchedQueries.exe.dir/utf_main.cpp.o.d -o test/CMakeFiles/ArborX_DetailsBatchedQueries.exe.dir/utf_main.cpp.o -c ../test/utf_main.cpp
nvcc fatal   : Unknown option 'fopenmp,-fexceptions,-pthread'    
```
I think `COMPILE_OPTIONS` cannot have spaces in it. In standalone Kokkos it's
```
  INTERFACE_COMPILE_OPTIONS "\$<\$<COMPILE_LANGUAGE:CXX>:>;\$<\$<COMPILE_LANGUAGE:CXX>:-expt-extended-lambda;-arch=sm_61>;\$<\$<COMPILE_LANGUAGE:CXX>:>;\$<\$<COMPILE_LANGUAGE:CXX>:-Xcompiler;-fopenmp>"                                                                                                            
```

**Comment by jjwilke** [comment]
Yeah, this is going to be fragile until we move to CMake 3.13.  Okay, I think I'm getting there.

**Comment by aprokop** [comment]
What's in CMake 3.13?

**Comment by jjwilke** [comment]
Flag deduplication stuff. You can declare a flag with a space "-Xcompiler -myflag".

**Comment by aprokop** [comment]
The newest version seems to work:
```CMake
IF (NOT TARGET Trilinos::kokkos)                                                                                                                                                                                                                                                                                     
  # Compute the installation prefix relative to this file.                                                                                                                                                                                                                                                           
  get_filename_component(KOKKOS_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)                                                                                                                                                                                                                                     
  get_filename_component(KOKKOS_IMPORT_PREFIX "${KOKKOS_IMPORT_PREFIX}" PATH)                                                                                                                                                                                                                                        
  get_filename_component(KOKKOS_IMPORT_PREFIX "${KOKKOS_IMPORT_PREFIX}" PATH)                                                                                                                                                                                                                                        
  get_filename_component(KOKKOS_IMPORT_PREFIX "${KOKKOS_IMPORT_PREFIX}" PATH)                                                                                                                                                                                                                                        
  if(KOKKOS_IMPORT_PREFIX STREQUAL "/")                                                                                                                                                                                                                                                                              
    set(KOKKOS_IMPORT_PREFIX "")                                                                                                                                                                                                                                                                                     
  endif()                                                                                                                                                                                                                                                                                                            
  add_library(Trilinos::kokkos INTERFACE IMPORTED)                                                                                                                                                                                                                                                                   
  set_target_properties(Trilinos::kokkos PROPERTIES                                                                                                                                                                                                                                                                  
    INTERFACE_LINK_LIBRARIES "kokkosalgorithms;kokkoscontainers;kokkoscore;-fopenmp"                                                                                                                                                                                                                                 
    INTERFACE_COMPILE_FEATURES "cxx_std_11"                                                                                                                                                                                                                                                                          
    INTERFACE_COMPILE_OPTIONS "$<$<COMPILE_LANGUAGE:CXX>:-expt-extended-lambda;-arch=sm_61;-Xcompiler;-fopenmp>"                                                                                                                                                                                                     
    INTERFACE_INCLUDE_DIRECTORIES "${KOKKOS_IMPORT_PREFIX}/include"                                                                                                                                                                                                                                                  
  )                                                                                                                                                                                                                                                                                                                  
ENDIF()   
```
I was successfully build, link and run all ArborX tests against the Trilinos installation.

**Comment by aprokop** [comment]
I also tried to create an alias
```CMake
  add_library(Kokkos::kokkos ALIAS Trilinos::kokkos)                                                                                                                                                                 
```
but that failed with
```
  add_library cannot create ALIAS target "Kokkos::kokkos" because target
  "Trilinos::kokkos" is imported but not globally visible.
```

**Comment by jjwilke** [comment]
Correct, you would not be able to do this unless I declared the Trilinos::kokkos `ALIAS GLOBAL`. I could do this, but I think you would only want the `Trilinos::kokkos` target visible in the scope the package was loaded.

Why do you need the alias? (I know that's the name you have been using).  The Kokkos standalone target is going to be named `Kokkos::kokkos` so I would think you would want the Trilinos target to be named something else (i.e. `Trilinos::kokkos`). Maybe I'm misunderstanding the use case.

**Comment by aprokop** [comment]
> Why do you need the alias?

I don't really need it. It would have removed couple lines of code of no significance.

> I would think you would want the Trilinos target to be named something else

From my point of view, I'd like to be able to
```
find_package(Kokkos REQUIRED)
target_link_libraries(blah INTERFACEKokkos::kokkos)
```
independent of whether it's a standalone or Trilinos installation. Having `Trilinos::kokkos` target means having extra code to deal with that. I would prefer not having that, but I'm not sure what the implications are.

**Comment by jjwilke** [comment]
Ah, okay. I see what you're saying. I misread the issue and thought you installed both Trilinos and standalone Kokkos and were using both (which would be really fragile when linking - so not sure why I though that's what you were saying).

Oooookaaaaay. I think the correct thing to do is change the name to `Kokkos::kokkos` rather than making it a global import.

**Comment by jjwilke** [comment]
Renamed. Should be ready.

File: cmake/KokkosTrilinosConfig.cmake.in [hunk 1]
```diff
@@ -0,0 +1,17 @@
+IF (NOT TARGET Trilinos::kokkos)
```
Comment (aprokop) [inline cmake/KokkosTrilinosConfig.cmake.in]: This should be `Kokkos::kokkos`, too.

Comment (jjwilke) [reply cmake/KokkosTrilinosConfig.cmake.in]: Yes it should!

**Comment by aprokop** [comment]
It seems to work fine. I tested:
- Serial+OpenMP+Cuda
  Have both openmp and cuda flags
- Serial+Cuda
  Have Cuda but no openmp flags
- Serial+OpenMP
  Have openmp flags but no CUDA

So, I think it's ready.

**Comment by jjwilke** [comment]
I will chat with @ndellingwood about doing a Trilinos promotion soon.

**Pull request merged**